src/HttpServer.php

```diff
@@ -60,50 +64,101 @@ public function addMiddleware($middleware)
-                });
-                $headerStream->emit('data', array($data));
-            }
+                if ($request->hasHeader('Content-Length')) {
+                    $contentLength = $request->getHeaderLine('Content-Length');
 
-            if (isset($request)) {
-                if ($that->isChunkedEncodingActive($request)) {
-                    $chunkedDecoder->on('data', function ($chunk) use (&$bodyBuffer, &$request, $that, $connection) {
-                        $bodyBuffer .= $chunk;
-                        if (strlen($chunk) == 0) {
-                            $that->sendBody($bodyBuffer, $connection, $request);
-                        }
-                    });
-                    $chunkStream->emit('data', array($data));
-                } else {
-                    $bodyBuffer .= $data;
-                    $contentLengthArray = $request->getHeader('Content-Length');
-
-                    if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                        $that->sendBody($bodyBuffer, $connection, $request);
-                    } else if (empty($contentLengthArray) || $contentLengthArray[0] == 0) {
-                        $that->handleRequest($connection, $request);
+                    $int = (int) $contentLength;
+                    if ((string)$int !== (string)$contentLength) {
+                        // Send 400 status code if the value of 'Content-Length'
+                        // is not an integer or is duplicated
+                        $that->sendResponse(new Response(400), $connection);
+                        return;
                     }
                 }
+
+                $that->handleBody($request, $connection);
+
+                // remove header from $data, only body is left
+                $data = (string)substr($data, strlen($fullHeader));
+                if ($data !== '') {
+                    $connection->emit('data', array($data));
+                }
+            }
+        };
+
+        $connection->on('data', $listener);
+    }
+
+    /** @internal */
+    public function handleBody(RequestInterface $request, ConnectionInterface $connection)
+    {
+        if ($this->isChunkedEncodingActive($request)) {
+            // Add ChunkedDecoder to stream
+            $chunkedDecoder = new ChunkedDecoder($connection);
+            $bodyStream = new HttpBodyStream($chunkedDecoder);
+            $request = $request->withBody($bodyStream);
+            $this->handleRequest($connection, $request);
+            return;
+        }
+
+        $bodyStream = new HttpBodyStream($connection);
+        $request = $request->withBody($bodyStream);
+        $this->handleRequest($connection, $request);
+
+        if (!$request->hasHeader('Content-Length')) {
+            // Request hasn't defined 'Content-Length' will ignore rest of the request
+            // and ends the stream
+            $bodyStream->emit('end', array());
```

Comment:
This is tested but had an error elsewhere. Currently keep-alive is not supported so the remaining data will currently ignored. I'm working on it but it will be handled in an seperated PR